core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java

```diff
@@ -0,0 +1,497 @@
+    public void updateAddresses(final List<EquivalentAddressGroup> newAddressGroups) {
+      Preconditions.checkNotNull(newAddressGroups, "newAddressGroups");
+      checkListHasNoNulls(newAddressGroups, "newAddressGroups contains null entry");
+      Preconditions.checkArgument(!newAddressGroups.isEmpty(), "newAddressGroups is empty");
+      final List<EquivalentAddressGroup> newImmutableAddressGroups =
+        Collections.unmodifiableList(new ArrayList<>(newAddressGroups));
+
+      helper.getSynchronizationContext().execute(new Runnable() {
+        @Override
+        public void run() {
+          SocketAddress previousAddress = addressIndex.getCurrentAddress();
+          addressIndex.updateGroups(newImmutableAddressGroups);
+          Set<SocketAddress> oldAddrs = new HashSet<>(subchannels.keySet());
+          Set<SocketAddress> newAddrs = new HashSet<>();
+          for (EquivalentAddressGroup endpoint : newImmutableAddressGroups) {
+            for (SocketAddress addr : endpoint.getAddresses()) {
+              newAddrs.add(addr);
+              if (!subchannels.containsKey(addr)) {
+                List<EquivalentAddressGroup> addrs = new ArrayList<>();
+                addrs.add(new EquivalentAddressGroup(addr));
+                final Subchannel subchannel = helper.createSubchannel(
+                    CreateSubchannelArgs.newBuilder()
+                    .setAddresses(addrs)
+                    .build());
+                subchannels.put(addr, subchannel);
+                subchannel.start(new SubchannelStateListener() {
+                  @Override
+                  public void onSubchannelState(ConnectivityStateInfo stateInfo) {
+                    processSubchannelState(subchannel, stateInfo);
+                  }
+                });
+              }
+            }
+          }
+
+          // remove old subchannels that were not in new address list
+          for (SocketAddress oldAddr : oldAddrs) {
+            if (!newAddrs.contains(oldAddr)) {
+              subchannels.get(oldAddr).shutdown();
+              subchannels.remove(oldAddr);
+            }
+          }
+
+          // start connection attempt at first address
+          if (addressIndex.seekTo(previousAddress)) {
+            // start connection attempt at first address
+            if (currentState == CONNECTING || currentState == TRANSIENT_FAILURE) {
+              addressIndex.reset();
+              requestConnection();
+            // TODO: verify desired behavior (ideally, we should restart at first address)
+            } else if (currentState == IDLE) {
+              SubchannelPicker picker = new RequestConnectionPicker(subchannels.get(addressIndex.getCurrentAddress()));
+              updateBalancingState(IDLE, picker);
+            }
+          } else {
+            // start connection attempt at first address
+            if (currentState == CONNECTING || currentState == TRANSIENT_FAILURE) {
+              requestConnection();
```

Comment:
I believe seekTo() only sets the indexes if it returns true (if it is able to find the address in the new group). So, the indexes would be unchanged if the statement returns false because of the continue statement. But, the updateGroups() in L105 should reset the index!